src/transform/src/semijoin_idempotence.rs

```diff
@@ -47,45 +69,108 @@ impl crate::Transform for SemijoinIdempotence {
         relation: &mut MirRelationExpr,
         _: TransformArgs,
     ) -> Result<(), crate::TransformError> {
-        {
-            // Iteratively descend the expression, at each node attempting to simplify a join.
-            let mut let_replacements = BTreeMap::<Id, Vec<Replacement>>::new();
-            let mut gets_behind_gets = BTreeMap::<Id, Vec<(Id, Vec<MirScalarExpr>)>>::new();
-            let mut worklist = vec![&mut *relation];
-            while let Some(expr) = worklist.pop() {
-                match expr {
-                    MirRelationExpr::Let { id, value, .. } => {
-                        let_replacements.insert(
-                            Id::Local(*id),
-                            list_replacements(&*value, &let_replacements, &gets_behind_gets),
-                        );
-                        gets_behind_gets
-                            .insert(Id::Local(*id), as_filtered_get(value, &gets_behind_gets));
+        let mut let_replacements = BTreeMap::<LocalId, Vec<Replacement>>::new();
+        let mut gets_behind_gets = BTreeMap::<LocalId, Vec<(Id, Vec<MirScalarExpr>)>>::new();
+        self.action(relation, &mut let_replacements, &mut gets_behind_gets)?;
+
+        mz_repr::explain::trace_plan(&*relation);
+        Ok(())
+    }
+}
+
+impl SemijoinIdempotence {
+    fn action(
+        &self,
+        expr: &mut MirRelationExpr,
+        let_replacements: &mut BTreeMap<LocalId, Vec<Replacement>>,
+        gets_behind_gets: &mut BTreeMap<LocalId, Vec<(Id, Vec<MirScalarExpr>)>>,
+    ) -> Result<(), crate::TransformError> {
+        // At each node, either gather info about Let bindings or attempt to simplify a join.
+        self.checked_recur(|_| {
```

Comment:
nit: `rust-analyzer` currently get confused because when you reference `match expr` in the `checked_recur` lambda the type inferred for `expr` is `&MirRelationExpr` and not `&mut MirRelationExpr`. 

Is it a problem to change this to `self.checked_recur(|_| {`?

Reply:
I'm not sure what should I change it to. `self.checked_recur(&mut |_| {` doesn't compile. Also, I can't experiment with potential fixes because I don't use VS Code.

Reply:
Sorry, I meant `self.checked_recur(move |_| {` instead of `self.checked_recur(|_| {` so the compiler _and_ `rust-analyzer` both can infer that the `expr: &mut MirRelationExpr` has moved.

Reply:
Ah, ok, that's working.